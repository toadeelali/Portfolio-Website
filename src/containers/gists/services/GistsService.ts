import {GistFilesModel, GistDOMModel, GistModel, GistsData} from '../models'
import {GistsApi} from './GistsApi'
import {GistsEnum} from '../models/GistsEnum'
import {gistLogosService} from './GistLogosService'
import {storageService} from '../../../shared/services/StorageService'

const REGEX_WEBSITE = /#web-featured$/
const REGEX_LOGO_DESCRIPTION = /^\[(.*)\]/

const transformData = (gists: GistModel[]): GistDOMModel[] => {
  return gists.map((gist: GistModel) => {
    const files: GistFilesModel[] = Object.values(gist.files)
    const title = gist.description.replace(REGEX_LOGO_DESCRIPTION, '').trim()
    const languageMatch = gist.description.match(REGEX_LOGO_DESCRIPTION)
    const language = languageMatch && languageMatch[1] && languageMatch[1].replace(REGEX_WEBSITE, '').trim()

    return {
      id: gist.id,
      url: gist.html_url,
      files: files,
      title: title || GistsEnum.title,
      language: language || files[0].language || GistsEnum.logo
    }
  })
}

const filterData = (gists: GistModel[]): GistModel[] => {
  const regex = new RegExp(REGEX_WEBSITE, 'gmi')

  return gists
    .filter((gist) => gist.public && gist.description.match(regex))
    .slice(0, GistsEnum.limit)
    .map((gist: GistModel) => ({
      ...gist,
      description: gist.description.replace(regex, '').trim()
    }))
}

const getSavedGists = (): GistsData | undefined => {
  const [savedGists, error] = storageService.getItem('gists') as GistsData[]

  if (error) return error
  if (!savedGists) return undefined

  return {
    date: Date.now(),
    collection: savedGists.collection,
    logos: savedGists.logos
  }
}

const shouldSaveGists = (gists: GistsData): boolean => {
  if (!gists || !gists.date) return false

  const hour = 1000 * 60 * 60
  const diff = (Date.now() - gists.date) / hour

  return Math.abs(Math.round(diff)) > 1
}

const saveGists = (gists: GistsData): GistsData | void => {
  const savedGists = getSavedGists()

  if (savedGists && !shouldSaveGists(savedGists)) return gists

  storageService.setItem('gists', JSON.stringify(gists))
}

const getGists = async (): Promise<GistsData> => {
  const savedGists = getSavedGists()

  if (savedGists) return savedGists

  const gists = await GistsApi.fetchGists()
  const filteredGists = filterData(gists.collection)

  return {
    date: Date.now(),
    collection: transformData(filteredGists),
    logos: gistLogosService.transformLogos(gists.logos)
  }
}

export const gistsService = {
  getGists,
  saveGists
}
