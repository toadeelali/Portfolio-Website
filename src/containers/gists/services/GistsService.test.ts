import fetchMock, {enableFetchMocks} from 'jest-fetch-mock'

import {gistsService} from './GistsService'

import {gistsResponseMock} from '../../../../test/mocks'

describe.skip('Gists Service', () => {
  beforeEach(() => {
    Object.defineProperty(window, 'localStorage', {
      value: {
        getItem: jest.fn(() => JSON.stringify(gistsResponseMock)),
        setItem: jest.fn(() => gistsResponseMock)
      },
      writable: true
    })

    enableFetchMocks()
  })

  afterEach(() => {
    fetchMock.resetMocks()
  })

  describe('Fetching Gists', () => {
    it('getGists: return fetched gists', async (done) => {
      fetchMock.mockResponseOnce(JSON.stringify(gistsResponseMock))

      const result = await gistsService.getGists()

      expect(result).toMatchSnapshot()
      done()
    })

    // it('getGists: return error when trying to fetch gists', async (done) => {
    //   fetchMock.mockRejectOnce(new Error('Not Found!'))

    //   const result = await gistsService.getGists()

    //   expect(result.toString()).toBe('Error: Not Found!')
    //   expect((fetch as any).mock.calls.length).toEqual(1)
    //   expect(result).toMatchSnapshot()
    //   done()
    // })
  })

  describe('Local Storage', () => {
    it('saveGists: should set gists in local storage', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          setItem: jest.fn(() => gistsResponseMock)
        }
      })

      const now = Date.now()
      gistsService.saveGists({
        date: now,
        logos: [],
        collection: [
          {
            id: '1',
            url: 'https://example.com',
            files: [
              {
                filename: 'test.js',
                type: 'application/javascript',
                language: 'JavaScript',
                raw_url: 'https://example.com/gist',
                size: 1000
              }
            ],
            title: 'Gist Sample',
            language: 'JavaScript'
          }
        ]
      })

      expect(window.localStorage.setItem).toHaveBeenCalledWith(
        'ADEEL-ALI:GISTS',
        `{"date":${now},"collection":[{"id":"1","url":"https://example.com","files":[{"filename":"test.js","type":"application/javascript","language":"JavaScript","raw_url":"https://example.com/gist","size":1000}],"title":"Gist Sample","language":"JavaScript"}]}`
      )
    })

    it('saveGists: should not call localStorage setItem', () => {
      Object.defineProperty(window, 'localStorage', {
        value: {
          setItem: jest.fn(() => gistsResponseMock)
        }
      })

      gistsService.saveGists(undefined as any)

      expect(window.localStorage.setItem).toHaveBeenCalledTimes(0)
    })
  })
})
