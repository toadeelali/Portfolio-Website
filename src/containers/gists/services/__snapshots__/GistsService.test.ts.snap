// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gists Service Fetching Gists getGists: return fetched gists 1`] = `
Object {
  "collection": Array [
    Object {
      "files": Array [
        Object {
          "filename": "intl-datetime-format.js",
          "language": "JavaScript",
          "raw_url": "https://gist.githubusercontent.com/wikz/6550840f43473d5b54d4e9232ba67ab0/raw/f6bdae2a990c20e5ac1578e90bc3cd77f4ce795e/intl-datetime-format.js",
          "size": 573,
          "type": "application/javascript",
        },
      ],
      "id": "6550840f43473d5b54d4e9232ba67ab0",
      "language": "JavaScript",
      "title": "Formatting date with Intl.DateTimeFormat",
      "url": "https://gist.github.com/6550840f43473d5b54d4e9232ba67ab0",
    },
    Object {
      "files": Array [
        Object {
          "filename": "dark-mode.css",
          "language": "CSS",
          "raw_url": "https://gist.githubusercontent.com/wikz/9e575e3a22813fdd3ab76aa33946a766/raw/ffba0089e8db90414cfc356935b1611afe6ebb52/dark-mode.css",
          "size": 613,
          "type": "text/css",
        },
      ],
      "id": "9e575e3a22813fdd3ab76aa33946a766",
      "language": "CSS",
      "title": "Easiest way to add Dark Mode",
      "url": "https://gist.github.com/9e575e3a22813fdd3ab76aa33946a766",
    },
    Object {
      "files": Array [
        Object {
          "filename": "git-clean-branches-alias.sh",
          "language": "Shell",
          "raw_url": "https://gist.githubusercontent.com/wikz/102ed3ae45494862c9b1381fb6749275/raw/81f54d34973ae1419b3fa5724b2571e08b5d17f5/git-clean-branches-alias.sh",
          "size": 306,
          "type": "application/x-sh",
        },
      ],
      "id": "102ed3ae45494862c9b1381fb6749275",
      "language": "Git",
      "title": "Clean Branches Alias",
      "url": "https://gist.github.com/102ed3ae45494862c9b1381fb6749275",
    },
    Object {
      "files": Array [
        Object {
          "filename": ".hyper.js",
          "language": "JavaScript",
          "raw_url": "https://gist.githubusercontent.com/wikz/2980bf771357f4caab71e8ae571f10c9/raw/e78cec2e2086dab2e96e766b401d08312a6e67a2/.hyper.js",
          "size": 5898,
          "type": "application/javascript",
        },
      ],
      "id": "2980bf771357f4caab71e8ae571f10c9",
      "language": "JavaScript",
      "title": "Terminal: My Hyper Configuration",
      "url": "https://gist.github.com/2980bf771357f4caab71e8ae571f10c9",
    },
    Object {
      "files": Array [
        Object {
          "filename": "get-latest-tag-on-git.sh",
          "language": "Shell",
          "raw_url": "https://gist.githubusercontent.com/wikz/5b5503c60df6371c22d611a0c87c2821/raw/486a95da7c14fbd3c5cc7801f273bf72a9a28ea8/get-latest-tag-on-git.sh",
          "size": 637,
          "type": "application/x-sh",
        },
      ],
      "id": "5b5503c60df6371c22d611a0c87c2821",
      "language": "Git",
      "title": "Getting the latest tag",
      "url": "https://gist.github.com/5b5503c60df6371c22d611a0c87c2821",
    },
  ],
  "date": 1592786022162,
  "logos": Array [],
}
`;
